### PR TITLE
Configurable Z Loss

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
### Overview

Following [https://github.com/pytorch/torchtitan/issues/2523](https://github.com/pytorch/torchtitan/issues/2523), this PR introduces an initial implementation of **z-loss** and includes some refactoring to make the loss configuration more flexible.

### Z-Loss

I added z-loss support to the cross-entropy loss. The implementation is inspired by the one used in OLMo:
[https://github.com/allenai/OLMo-core/blob/main/src/olmo_core/nn/cross_entropy_loss.py](https://github.com/allenai/OLMo-core/blob/main/src/olmo_core/nn/cross_entropy_loss.py).

If `z_loss_weight` is set to a value different from `0`, the z-loss is computed and added to the cross-entropy loss, scaled by `z_loss_weight`.

### Refactoring

I also refactored the loss configuration so it can be defined via the CLI or through TOML configuration files. This is just a proposal and can be adapted if it does not align with the Torchtitan design.

The current setup allows configuring multiple loss types (currently **MSE** and **CrossEntropy**). Each loss is defined as a configurable object with the following fields:

* **enable**: whether the loss is active (exactly one loss must be enabled)
* **compile**: if `true`, the loss module is compiled with `torch.compile`

Additional options are available for specific losses:

* **CrossEntropyLoss**

  * `z_loss_weight` (default: `0.0`)
  * `ignore_index` (default: `-100`)

Both **CrossEntropy** and **MSE** return a `LossOutput` object containing:

* `main`: the loss used for gradient computation (the one `.backward()` should be called on)
* `aux`: a dictionary containing auxiliary values intended only for logging

For example, the CrossEntropy loss populates `LossOutput.aux` with both the **unscaled z-loss** and the **raw cross-entropy loss**.

### CODE REVIEW

File: torchtitan/protocols/model_spec.py [hunk 1]
```diff
@@ -39,7 +39,6 @@ class ModelSpec:
     # resolver does not handle ``Callable[..., X]`` (Ellipsis as param spec).
     # The detailed TypeAliases above are still available for use in function
     # signatures elsewhere in the codebase.
-    build_loss_fn: Callable
```
Comment (tianyu-l) [inline torchtitan/protocols/model_spec.py:42]: Changing loss configuration from "build function" to config-based sounds good. However,
- why do we move it from `ModelSpec` to `Trainer`?
- by default only CrossEntropyLoss is used, which changes the default for FLUX model

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -5,49 +5,132 @@
...
+    def mse_loss(pred: torch.Tensor, labels: torch.Tensor) -> LossOutput:
+        """Common MSE loss function with sum reduction for Transformer models training."""
+        return LossOutput(
+            main=torch.nn.functional.mse_loss(
+                pred.float(), labels.float().detach(), reduction="sum"
+            )
+        )
+
+
+class Loss(Configurable):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:110]: We should have one `BaseLoss` class which crossentropy and mse inherit, rather than one monolithic Loss class responsible for everything.

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -5,49 +5,132 @@
...
-def cross_entropy_loss(pred: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
-    """Cross-entropy loss with sum reduction for token-based normalization."""
-    return torch.nn.functional.cross_entropy(
-        pred.flatten(0, 1).float(),
-        labels.flatten(0, 1),
-        reduction="sum",
-        ignore_index=IGNORE_INDEX,
-    )
+    main: torch.Tensor
+    aux: dict[str, torch.Tensor] = field(default_factory=dict)
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:30]: It seems `aux` is only for logging. I think it'd be cleaner to wait until https://github.com/pytorch/torchtitan/pull/2607 lands.

Comment (francesco-bertolotti) [reply torchtitan/components/loss.py:30]: Yep aux is meant only for logging. No problem in waiting for https://github.com/pytorch/torchtitan/pull/2607

File: torchtitan/components/loss.py (hunk 2 above)
Comment (francesco-bertolotti) [reply torchtitan/components/loss.py:110]: Right now, cross-entropy and MSE losses are implemented as separate classes (though they could definitely share a common `BaseLoss` parent). The current `Loss` object simply routes to the appropriate implementation based on the configuration.

For example, the CLI for configuring the loss looks like this:

```bash
torchtitan/train.py \
  --config qwen3_debugmodel \
  --module qwen3 \
  --loss.cross-entropy.enable \
  --loss.cross-entropy.compile \
  --loss.cross-entropy.ignore-index -100 \
  --loss.cross-entropy.z-loss-weight 1e-5
```

File: torchtitan/protocols/model_spec.py (hunk 1 above)
Comment (francesco-bertolotti) [reply torchtitan/protocols/model_spec.py:42]: I was under the impression that this was no longer necessary, since the loss can now be configured directly from the config file. So I simplified it. However, it may make more sense to keep it, I'm happy to revert this change if needed.

File: torchtitan/components/loss.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/components/loss.py:110]: the current thinking around CLI is that
- python config is strictly closer to the actual object than CLI, thus more "minimal" to support
- the design should focus on this minimal nature, not biased (too much) by CLI concern
- users can always build CLI oriented config on top of existing config
